<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>hocort.pipelines.hisat2 API documentation</title>
<meta name="description" content="" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>hocort.pipelines.hisat2</code></h1>
</header>
<section id="section-intro">
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">import time
import os

from hocort.pipelines.pipeline import Pipeline
from hocort.aligners.hisat2 import HISAT2 as hs2
from hocort.parse.sam import SAM
from hocort.parser import ArgParser
import hocort.execute as exe


class HISAT2(Pipeline):
    &#34;&#34;&#34;
    HISAT2 pipeline which maps reads to a genome and includes/excludes matching reads from the output FastQ file/-s.

    &#34;&#34;&#34;
    def __init__(self):
        &#34;&#34;&#34;
        Sets the logger file source filename.

        Returns
        -------
        None

        &#34;&#34;&#34;
        super().__init__(__file__)

    def run(self, idx, seq1, out1, seq2=None, out2=None, hcfilter=False, threads=1, options=[]):
        &#34;&#34;&#34;
        Run function which starts the pipeline.

        Parameters
        ----------
        idx : string
            Path where the index is located.
        seq1 : string
            Path where the first input FastQ file is located.
        out1 : string
            Path where the first output FastQ file will be written.
        seq2 : string
            Path where the second input FastQ file is located.
        out2 : string
            Path where the second output FastQ file will be written.
        hcfilter : bool
            Whether to exclude or include the matching sequences from the output files.
        threads : int
            Number of threads to use.
        options : list
            An options list where additional arguments may be specified.

        Returns
        -------
        returncode : int
            Resulting returncode after the process is finished.

        &#34;&#34;&#34;
        self.debug_log_args(self.run.__name__, locals())
        if seq2 and not out2: return 1

        self.logger.warning(f&#39;Running pipeline: {self.__class__.__name__}&#39;)
        start_time = time.time()

        if len(options) &gt; 0:
            options = options
        else:
            options = [&#39;--sensitive&#39;, &#39;--sp 3,2&#39;, &#39;--mp 5,1&#39;]

        hs2_cmd = hs2().align(idx, seq1, seq2=seq2, threads=threads, options=options)
        if hs2_cmd == None: return 1
        fastq_cmd = SAM.sam_to_fastq(out1=out1, out2=out2, threads=threads, hcfilter=hcfilter)

        returncodes = exe.execute(hs2_cmd + fastq_cmd, pipe=True)

        self.logger.debug(returncodes)
        for returncode in returncodes:
            if returncode != 0: return 1

        end_time = time.time()
        self.logger.warning(f&#39;Pipeline {self.__class__.__name__} run time: {end_time - start_time} seconds&#39;)
        return 0

    def interface(self, args):
        &#34;&#34;&#34;
        Main function for the user interface. Parses arguments and starts the pipeline.

        Parameters
        ----------
        args : list
            This list is parsed by ArgumentParser.

        Returns
        -------
        None

        &#34;&#34;&#34;
        parser = ArgParser(
            description=f&#39;{self.__class__.__name__} pipeline&#39;,
            usage=f&#39;hocort map {self.__class__.__name__} [-h] [--threads &lt;int&gt;] [--host-contam-filter &lt;bool&gt;] -x &lt;idx&gt; -i &lt;fastq_1&gt; [&lt;fastq_2&gt;] -o &lt;fastq_1&gt; [&lt;fastq_2&gt;]&#39;
        )
        parser.add_argument(
            &#39;-x&#39;,
            &#39;--index&#39;,
            required=True,
            type=str,
            metavar=(&#39;&lt;idx&gt;&#39;),
            help=&#39;str: path to HISAT2 index (required)&#39;
        )
        parser.add_argument(
            &#39;-i&#39;,
            &#39;--input&#39;,
            required=True,
            type=str,
            nargs=(&#39;+&#39;),
            metavar=(&#39;&lt;fastq_1&gt;&#39;, &#39;&lt;fastq_2&gt;&#39;),
            help=&#39;str: path to sequence files, max 2 (.gz compression supported) (required)&#39;
        )
        parser.add_argument(
            &#39;-o&#39;,
            &#39;--output&#39;,
            required=True,
            type=str,
            nargs=(&#39;+&#39;),
            metavar=(&#39;&lt;fastq_1&gt;&#39;, &#39;&lt;fastq_2&gt;&#39;),
            help=&#39;str: path to output files, max 2 (.gz compression supported) (required)&#39;
        )
        parser.add_argument(
            &#39;-t&#39;,
            &#39;--threads&#39;,
            required=False,
            type=int,
            metavar=(&#39;&lt;int&gt;&#39;),
            default=os.cpu_count(),
            help=&#39;int: number of threads (default: max available on machine)&#39;
        )
        parser.add_argument(
            &#39;-f&#39;,
            &#39;--host-contam-filter&#39;,
            choices=[&#39;True&#39;, &#39;False&#39;],
            default=&#39;False&#39;,
            help=&#39;str: set to True to keep host sequences, False to keep everything besides host sequences (default: False)&#39;
        )
        parsed = parser.parse_args(args=args)

        idx = parsed.index
        seq = parsed.input
        out = parsed.output
        threads = parsed.threads if parsed.threads else 1
        hcfilter = True if parsed.host_contam_filter == &#39;True&#39; else False

        seq1 = seq[0]
        seq2 = None if len(seq) &lt; 2 else seq[1]
        out1 = out[0]
        out2 = None if len(out) &lt; 2 else out[1]

        return self.run(idx, seq1, out1, out2=out2, seq2=seq2, hcfilter=hcfilter, threads=threads)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="hocort.pipelines.hisat2.HISAT2"><code class="flex name class">
<span>class <span class="ident">HISAT2</span></span>
</code></dt>
<dd>
<div class="desc"><p>HISAT2 pipeline which maps reads to a genome and includes/excludes matching reads from the output FastQ file/-s.</p>
<p>Sets the logger file source filename.</p>
<h2 id="returns">Returns</h2>
<dl>
<dt><code>None</code></dt>
<dd>&nbsp;</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class HISAT2(Pipeline):
    &#34;&#34;&#34;
    HISAT2 pipeline which maps reads to a genome and includes/excludes matching reads from the output FastQ file/-s.

    &#34;&#34;&#34;
    def __init__(self):
        &#34;&#34;&#34;
        Sets the logger file source filename.

        Returns
        -------
        None

        &#34;&#34;&#34;
        super().__init__(__file__)

    def run(self, idx, seq1, out1, seq2=None, out2=None, hcfilter=False, threads=1, options=[]):
        &#34;&#34;&#34;
        Run function which starts the pipeline.

        Parameters
        ----------
        idx : string
            Path where the index is located.
        seq1 : string
            Path where the first input FastQ file is located.
        out1 : string
            Path where the first output FastQ file will be written.
        seq2 : string
            Path where the second input FastQ file is located.
        out2 : string
            Path where the second output FastQ file will be written.
        hcfilter : bool
            Whether to exclude or include the matching sequences from the output files.
        threads : int
            Number of threads to use.
        options : list
            An options list where additional arguments may be specified.

        Returns
        -------
        returncode : int
            Resulting returncode after the process is finished.

        &#34;&#34;&#34;
        self.debug_log_args(self.run.__name__, locals())
        if seq2 and not out2: return 1

        self.logger.warning(f&#39;Running pipeline: {self.__class__.__name__}&#39;)
        start_time = time.time()

        if len(options) &gt; 0:
            options = options
        else:
            options = [&#39;--sensitive&#39;, &#39;--sp 3,2&#39;, &#39;--mp 5,1&#39;]

        hs2_cmd = hs2().align(idx, seq1, seq2=seq2, threads=threads, options=options)
        if hs2_cmd == None: return 1
        fastq_cmd = SAM.sam_to_fastq(out1=out1, out2=out2, threads=threads, hcfilter=hcfilter)

        returncodes = exe.execute(hs2_cmd + fastq_cmd, pipe=True)

        self.logger.debug(returncodes)
        for returncode in returncodes:
            if returncode != 0: return 1

        end_time = time.time()
        self.logger.warning(f&#39;Pipeline {self.__class__.__name__} run time: {end_time - start_time} seconds&#39;)
        return 0

    def interface(self, args):
        &#34;&#34;&#34;
        Main function for the user interface. Parses arguments and starts the pipeline.

        Parameters
        ----------
        args : list
            This list is parsed by ArgumentParser.

        Returns
        -------
        None

        &#34;&#34;&#34;
        parser = ArgParser(
            description=f&#39;{self.__class__.__name__} pipeline&#39;,
            usage=f&#39;hocort map {self.__class__.__name__} [-h] [--threads &lt;int&gt;] [--host-contam-filter &lt;bool&gt;] -x &lt;idx&gt; -i &lt;fastq_1&gt; [&lt;fastq_2&gt;] -o &lt;fastq_1&gt; [&lt;fastq_2&gt;]&#39;
        )
        parser.add_argument(
            &#39;-x&#39;,
            &#39;--index&#39;,
            required=True,
            type=str,
            metavar=(&#39;&lt;idx&gt;&#39;),
            help=&#39;str: path to HISAT2 index (required)&#39;
        )
        parser.add_argument(
            &#39;-i&#39;,
            &#39;--input&#39;,
            required=True,
            type=str,
            nargs=(&#39;+&#39;),
            metavar=(&#39;&lt;fastq_1&gt;&#39;, &#39;&lt;fastq_2&gt;&#39;),
            help=&#39;str: path to sequence files, max 2 (.gz compression supported) (required)&#39;
        )
        parser.add_argument(
            &#39;-o&#39;,
            &#39;--output&#39;,
            required=True,
            type=str,
            nargs=(&#39;+&#39;),
            metavar=(&#39;&lt;fastq_1&gt;&#39;, &#39;&lt;fastq_2&gt;&#39;),
            help=&#39;str: path to output files, max 2 (.gz compression supported) (required)&#39;
        )
        parser.add_argument(
            &#39;-t&#39;,
            &#39;--threads&#39;,
            required=False,
            type=int,
            metavar=(&#39;&lt;int&gt;&#39;),
            default=os.cpu_count(),
            help=&#39;int: number of threads (default: max available on machine)&#39;
        )
        parser.add_argument(
            &#39;-f&#39;,
            &#39;--host-contam-filter&#39;,
            choices=[&#39;True&#39;, &#39;False&#39;],
            default=&#39;False&#39;,
            help=&#39;str: set to True to keep host sequences, False to keep everything besides host sequences (default: False)&#39;
        )
        parsed = parser.parse_args(args=args)

        idx = parsed.index
        seq = parsed.input
        out = parsed.output
        threads = parsed.threads if parsed.threads else 1
        hcfilter = True if parsed.host_contam_filter == &#39;True&#39; else False

        seq1 = seq[0]
        seq2 = None if len(seq) &lt; 2 else seq[1]
        out1 = out[0]
        out2 = None if len(out) &lt; 2 else out[1]

        return self.run(idx, seq1, out1, out2=out2, seq2=seq2, hcfilter=hcfilter, threads=threads)</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="hocort.pipelines.pipeline.Pipeline" href="pipeline.html#hocort.pipelines.pipeline.Pipeline">Pipeline</a></li>
<li>abc.ABC</li>
</ul>
<h3>Methods</h3>
<dl>
<dt id="hocort.pipelines.hisat2.HISAT2.run"><code class="name flex">
<span>def <span class="ident">run</span></span>(<span>self, idx, seq1, out1, seq2=None, out2=None, hcfilter=False, threads=1, options=[])</span>
</code></dt>
<dd>
<div class="desc"><p>Run function which starts the pipeline.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>idx</code></strong> :&ensp;<code>string</code></dt>
<dd>Path where the index is located.</dd>
<dt><strong><code>seq1</code></strong> :&ensp;<code>string</code></dt>
<dd>Path where the first input FastQ file is located.</dd>
<dt><strong><code>out1</code></strong> :&ensp;<code>string</code></dt>
<dd>Path where the first output FastQ file will be written.</dd>
<dt><strong><code>seq2</code></strong> :&ensp;<code>string</code></dt>
<dd>Path where the second input FastQ file is located.</dd>
<dt><strong><code>out2</code></strong> :&ensp;<code>string</code></dt>
<dd>Path where the second output FastQ file will be written.</dd>
<dt><strong><code>hcfilter</code></strong> :&ensp;<code>bool</code></dt>
<dd>Whether to exclude or include the matching sequences from the output files.</dd>
<dt><strong><code>threads</code></strong> :&ensp;<code>int</code></dt>
<dd>Number of threads to use.</dd>
<dt><strong><code>options</code></strong> :&ensp;<code>list</code></dt>
<dd>An options list where additional arguments may be specified.</dd>
</dl>
<h2 id="returns">Returns</h2>
<dl>
<dt><strong><code>returncode</code></strong> :&ensp;<code>int</code></dt>
<dd>Resulting returncode after the process is finished.</dd>
</dl></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def run(self, idx, seq1, out1, seq2=None, out2=None, hcfilter=False, threads=1, options=[]):
    &#34;&#34;&#34;
    Run function which starts the pipeline.

    Parameters
    ----------
    idx : string
        Path where the index is located.
    seq1 : string
        Path where the first input FastQ file is located.
    out1 : string
        Path where the first output FastQ file will be written.
    seq2 : string
        Path where the second input FastQ file is located.
    out2 : string
        Path where the second output FastQ file will be written.
    hcfilter : bool
        Whether to exclude or include the matching sequences from the output files.
    threads : int
        Number of threads to use.
    options : list
        An options list where additional arguments may be specified.

    Returns
    -------
    returncode : int
        Resulting returncode after the process is finished.

    &#34;&#34;&#34;
    self.debug_log_args(self.run.__name__, locals())
    if seq2 and not out2: return 1

    self.logger.warning(f&#39;Running pipeline: {self.__class__.__name__}&#39;)
    start_time = time.time()

    if len(options) &gt; 0:
        options = options
    else:
        options = [&#39;--sensitive&#39;, &#39;--sp 3,2&#39;, &#39;--mp 5,1&#39;]

    hs2_cmd = hs2().align(idx, seq1, seq2=seq2, threads=threads, options=options)
    if hs2_cmd == None: return 1
    fastq_cmd = SAM.sam_to_fastq(out1=out1, out2=out2, threads=threads, hcfilter=hcfilter)

    returncodes = exe.execute(hs2_cmd + fastq_cmd, pipe=True)

    self.logger.debug(returncodes)
    for returncode in returncodes:
        if returncode != 0: return 1

    end_time = time.time()
    self.logger.warning(f&#39;Pipeline {self.__class__.__name__} run time: {end_time - start_time} seconds&#39;)
    return 0</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="hocort.pipelines.pipeline.Pipeline" href="pipeline.html#hocort.pipelines.pipeline.Pipeline">Pipeline</a></b></code>:
<ul class="hlist">
<li><code><a title="hocort.pipelines.pipeline.Pipeline.debug_log_args" href="pipeline.html#hocort.pipelines.pipeline.Pipeline.debug_log_args">debug_log_args</a></code></li>
<li><code><a title="hocort.pipelines.pipeline.Pipeline.interface" href="pipeline.html#hocort.pipelines.pipeline.Pipeline.interface">interface</a></code></li>
</ul>
</li>
</ul>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="hocort.pipelines" href="index.html">hocort.pipelines</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="hocort.pipelines.hisat2.HISAT2" href="#hocort.pipelines.hisat2.HISAT2">HISAT2</a></code></h4>
<ul class="">
<li><code><a title="hocort.pipelines.hisat2.HISAT2.run" href="#hocort.pipelines.hisat2.HISAT2.run">run</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>